### ISSUE TITLE
Feature request: debounce triggering auto-completion

### ISSUE THREAD

**Issue by mystilleef** [hrsh7th/nvim-cmp#598]
The older `nvim-compe` had these options, and I loved them. In fact, I resisted switching to `nvim-cmp` for so long because I had tweaked the completion behavior to perfection and to be less “annoying” using these options. Can you bring this options back?

**Comment by hrsh7th** [comment]
Please explain how annoying with current setting.

**Comment by hrsh7th** [comment]
Isn't the `keyword length` option enough?

Sorry. I am careful about exposing the options as public.
I can add this but I want to know the rational reason.

**Comment by mystilleef** [comment]
> Isn't the `keyword length` option enough?
> 

No, because I use `cmp` to trigger local snippets, which I prioritize above all sources. 

The behavior I want is that when I'm typing, I don't want the completion window updating with new items from sources until I stop typing for 1.5 seconds. Setting `THROTTLE_TIME=1500` and `SOURCE_TIMEOUT=100` helps me achieve this.

This for me creates  a less distracting and more predictable experience. It gives me a little control over when and how the completion lists gets populated. It also gives `tabnine` enough time to send me useful completions.

I find the default behavior of `cmp` too aggressive and noisy. It updates the list too quickly and sometimes in the middle of me trying to select an item it randomly inserts a new item.

> Sorry. I am careful about exposing the options as public. I can add this but I want to know the rational reason.

Understood.

**Comment by hrsh7th** [comment]
I'd tried `THROTTLE_TIME=1500` and `SOURCE_TIMEOUT=100`. It will cause the strange behavior...

**Comment by hrsh7th** [comment]
I hope to know,the ideal behaivior.

**Comment by mystilleef** [comment]
> 

Hmm, interesting. Works perfectly for me. I can see how `SOURCE_TIMEOUT` can cause issues for slow `lsp` servers. But I'm perfectly fine ignoring slow servers. 

The behavior I want is to delay querying the sources for new items until I stop typing. In other words, when I'm typing, I don't want any query requests sent to completion sources. Requests can be sent after I stop typing, but not when I'm typing.

This is so that the completion windows doesn't abruptly add new items when I'm typing. This is what creates the “distracting noise” for me. A delay of around 1.5 seconds is perfect for me, especially with `tabnine`.

I thought `THROTLLE_TIME` achieved this, but I haven't really studied the code. So my assumptions may be wrong.

In summary, I want an option that only updates the completion items when I'm not typing. When I'm typing, just filter what's already available in the completion window. Only add new items to the completion window after I stop typing, if it makes sense. 

I hope you understand what I'm trying to do.

**Comment by hrsh7th** [comment]
Therefore, the "debounce" function is required.
Hmm ... for now, I'll provide a workaround for debounce.

It seems that we have to think continuously.

**Comment by mystilleef** [comment]
> 

`DEBOUNCE` is the word I'm looking for. 

Yes, a `debounce` option when typing will go a long way. Especially since some `lsp` sources just love to through a lot of irrelevant items at you. I had to set `max_item_count = 5` because some `lsp` servers were driving me crazy. Thanks for that option, by the way, it's a lifesaver. :-)

I use a combination of `THROTTLE_TIME` and `SOURCE_TIMEOUT` to mimic `debouncing`. I've settled on 1000 and 200 respectively, and so far so good.

**Comment by hrsh7th** [comment]
The workaround.
```
lua << EOF
local cmp = require "cmp"
cmp.setup {
  completion = {
    autocomplete = false,
  },
  snippet = {
    expand = function(args)
      vim.fn["vsnip#anonymous"](args.body)
    end,
  },

  mapping = {
    ['<CR>'] = cmp.mapping.confirm({ select = true }),
  },

  sources = {
    { name = "buffer" },
  },
}
EOF

autocmd TextChangedI * call s:on_text_changed()
let s:timer_id = 0
function! s:on_text_changed() abort
  function! s:invoke() abort closure
lua << EOF
    local cmp = require('cmp')
    cmp.complete({ reason = cmp.ContextReason.Auto })
EOF
  endfunction
  call timer_stop(s:timer_id)
  let s:timer_id = timer_start(800, { -> s:invoke() })
endfunction
```

**Comment by hrsh7th** [comment]
https://github.com/hrsh7th/nvim-cmp/issues/435
https://github.com/hrsh7th/nvim-cmp/issues/266

**Comment by mystilleef** [comment]
For the benefit of those who want the workaround in `lua`, you can use the example below. You can place `debounce.lua` in your `lua` folder and call it from your `cmp` configuration.

```lua
-- debounce.lua

local M = {}

local cmp = require("cmp")
local timer = vim.loop.new_timer()

local DEBOUNCE_DELAY = 1000

function M.debounce()
  timer:stop()
  timer:start(
    DEBOUNCE_DELAY,
    0,
    vim.schedule_wrap(function()
      cmp.complete({ reason = cmp.ContextReason.Auto })
    end)
  )
end

return M
```

You can experiment with `DEBOUNCE_DELAY` to see what works for you.

Next, place the following snippet after `cmp.setup(...)`.

```lua

-- NOTE: change "plugin.cmp.debounce" to location of debounce.lua
-- If it's in the lua folder, then change to "debounce"

vim.cmd([[
  augroup CmpDebounceAuGroup
    au!
    au TextChangedI * lua require("plugin.cmp.debounce").debounce()
  augroup end
]])
```

**Comment by Aaronik** [comment]
Are you planning on adding a config option here? I would love a debounce just mentally. It's taxing and distracting for me to see the window constantly popping up and shifting around. I find myself having to look away from the screen when I'm typing so it's easier to hold focus on what I'm thinking about.

I'm currently trying the steps above but haven't yet got it working, though I'm confident I will be able to. But if this were just a simple config option, that'd have made my afternoon a lot easier :)

I hope you consider (or continue to support?) adding a flag for it. Thank you!

**Comment by Odas0R** [comment]
> For the benefit of those who want the workaround in `lua`, you can use the example below. You can place `debounce.lua` in your `lua` folder and call it from your `cmp` configuration.
> 
> ```lua
> -- debounce.lua
> 
> local M = {}
> 
> local cmp = require("cmp")
> local timer = vim.loop.new_timer()
> 
> local DEBOUNCE_DELAY = 1000
> 
> function M.debounce()
>   timer:stop()
>   timer:start(
>     DEBOUNCE_DELAY,
>     0,
>     vim.schedule_wrap(function()
>       cmp.complete({ reason = cmp.ContextReason.Auto })
>     end)
>   )
> end
> 
> return M
> ```
> 
> You can experiment with `DEBOUNCE_DELAY` to see what works for you.
> 
> Next, place the following snippet after `cmp.setup(...)`.
> 
> ```lua
> -- NOTE: change "plugin.cmp.debounce" to location of debounce.lua
> -- If it's in the lua folder, then change to "debounce"
> 
> vim.cmd([[
>   augroup CmpDebounceAuGroup
>     au!
>     au TextChangedI * lua require("plugin.cmp.debounce").debounce()
>   augroup end
> ]])
> ```
> 
> Also, don't forget to set `autocomplete = false` in `cmp.setup(...)`.
> 
> ```lua
> local cmp = require "cmp"
> cmp.setup {
>   ...
>   completion = {
>     autocomplete = false,
>     ...
>   },
> ...
> ```

Incredible performance booster. I thought that nvim-cmp had already a debounce mechanism behind the scenes???

Anyhow, thank you, thank you!

**Comment by gegoune** [comment]
This is different, it postpones request to cmp for completion.

**Comment by Odas0R** [comment]
> This is different, it postpones request to cmp for completion.

Yes, understood.

I didn't dig into the code so I don't have a picture of how it's built, but the performance issue was real.

This really helped me.

I have a light config, not many plugins, and I was having serious problems when coding on medium-large _typescript_ codebases...

**Comment by hrsh7th** [comment]
Now, We have `performance` configuration section. So I can add this feature.

Rough sketch.
```lua
cmp.setup {
  performance = {
    trigger_debounce_time = 500
  }
}
```

**Comment by ryuheechul** [comment]
The reason for me wanting this feature was the lags while typing continuously with suggestions kept coming up - due to spell source - which probably why other people want this feature as well. I just stumbled upon a fix for me which was setting `nolazyredraw` as below.
```lua
vim.o.lazyredraw = false -- or
vim.cmd [[ set nolazyredraw ]]
```

_So mine had been set to lazyredraw for some reason without me noticing and now the lag is gone  (at least the most of it) by explicit `vim.om.lazyredraw = false`._

**Comment by JoseConseco** [comment]
> For the benefit of those who want the workaround in `lua`, you can use the example below. You can place `debounce.lua` in your `lua` folder and call it from your `cmp` configuration.
> 
> ```lua
> -- debounce.lua
> 
> local M = {}
> 
> local cmp = require("cmp")
> local timer = vim.loop.new_timer()
> 
> local DEBOUNCE_DELAY = 1000
> 
> function M.debounce()
>   timer:stop()
>   timer:start(
>     DEBOUNCE_DELAY,
>     0,
>     vim.schedule_wrap(function()
>       cmp.complete({ reason = cmp.ContextReason.Auto })
>     end)
>   )
> end
> 
> return M
> ```
That is awesome. I jut noticed what cmp popup wont show if you for example use  `o` - create new line below, and go into insert mode.  IT seems ahat use   `cmp.ContextReason.Auto` will ignore empty lines  (like one created with `o`), thus I created slightly different condition:
```lua
  if not cmp.visible() and vim.fn.mode() == 'i' then
    -- Auto takes keyword_length into account, (wont ignore existing letters - unlike manual) but wont show anything if no letter in line. Good when line is not empty.
    -- Manual will show popup even if no letter, but then if user types few leters - the get  ignored. Good for empty lines.
    -- check if current line has any letters : if not use Manual
    local current_line = vim.api.nvim_get_current_line()
    local has_letters = string.find(current_line, "%a") -- "%a" matches any alphabet character (equivalent to [a-zA-Z])
    if has_letters then
      cmp.complete({ reason = cmp.ContextReason.Auto }) -- without reason, it will ignore letters before cursor...
    else
      cmp.complete({ reason = cmp.ContextReason.Manual }) -- without reason, it will ignore letters before cursor...
    end

    -- cmp.complete()  -- gives 'old' suggestion (like it does not see characters types after first bounce... even thought using bounce_trailing)
  end

```

**Comment by hb0nes** [comment]
Do we still need to use the workaround? The constant jumping around of boxes all over my editor makes me very distracted. Sometimes I can't see variable names from a function while typing.

**Comment by paride** [comment]
@hb0nes is that actually working for you? I have
```lua
    performance = {
      debounce = 1000,
    },
```
but I still get completion boxes jumping around while I type. See also https://github.com/hrsh7th/nvim-cmp/issues/2125.

**Comment by hb0nes** [comment]
Can't check right now but it definitely was working. Could also take a look
at blink.cmp, see if it support your use case.
